File: 141. Linked List Cycle/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
bool hasCycle(struct ListNode *head) {
    if (head == NULL) {
        return false;
    }
    
    struct ListNode *slowRunner = head->next;
    if (slowRunner == NULL) {
        return false;
    }
    struct ListNode *fastRunner = slowRunner->next;
    if (fastRunner == NULL) {
        return false;
    }    
    
    do {
        if (slowRunner == fastRunner) {
            return true;
        }
        
        slowRunner = slowRunner->next;
        fastRunner = fastRunner->next;
        if (fastRunner == NULL) {
            return false;
        }
        fastRunner = fastRunner->next;
    } while(fastRunner != NULL);
    return false;
}

File: 141. Linked List Cycle/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        
        if not head or not head.next:
            return False
        
        sptr = head
        fptr = head.next
        
        while sptr and fptr and fptr.next:
            if sptr == fptr:
                return True
            
            sptr = sptr.next
            fptr = fptr.next.next
            
        return False
